service/tablet_allocator.cc

```diff
@@ -1468,8 +1468,7 @@ class load_balancer {
             auto src_host = nodes_by_load.back();
             auto& src_node_info = nodes[src_host];
 
-            bool drain_skipped = src_node_info.shards_by_load.empty() && src_node_info.drained
-                    && !src_node_info.skipped_candidates.empty();
+            bool drain_skipped = src_node_info.shards_by_load.empty() && !src_node_info.skipped_candidates.empty();
```

Comment:
I think this can break per-table balance, since pick_candidate() assumes that if "drain_skipped" then we must move this tablet and there's no point looking for a better candidate. Which is true when node is drained, but not true during regular balancing:

```
            // Consider better alternatives.
            if (drain_skipped) {
                auto source_tablet = src_node_info.skipped_candidates.back().tablet;
                auto badness = evaluate_src_badness(nodes, source_tablet.table, src);
                co_await evaluate_targets(source_tablet, src, badness);
            } else {
                // Find a better candidate.
                // Consider different tables. For each table, first find the best source shard.
                // Then find the best target node. Then find the best shard on the target node.
                for (auto [table, load] : src_node_info.tablet_count_per_table) {

```

Reply:
Please note we only set drain_skipped to true, when src_node_info.shards_by_load is empty. And if it's empty, there won't be any candidate to be evaluated in the else branch, therefore we only drain skipped candidates once we're done with better candidates first (not in the skipped list).

see:
```
                for (auto [table, load] : src_node_info.tablet_count_per_table) {
                    migration_badness min_src_badness;
                    std::optional<tablet_replica> min_src;

                    ...

                    for (auto new_src_shard: src_node_info.shards_by_load) {
                        ...
                    }

                    if (!min_src) {
                        lblogger.debug("No candidates for table {} on {}", table, src.host);
                        continue;
                    }
```

But I suppose you're concerned about increasing table load on a node as a result of this change?

Note that one node has 10 tablets, whereas other has 8 tablets. Perhaps in the 1st node, it's 5 tablets of 1 table, and 5 of another. And 2nd node, it's 4 tablets of 1 table, and 4 of another. Maybe, table aware balancing considers this scenario as converged (4,5 or 5,4 doesn't matter), since it's only making the decision with one table in mind.

Reply:
My concern is that we pick tablets to move among skipped candidates in fifo order, while there could be better picks among skipped candidates themselves. First skiped candidate could break the per-table balance, while second one may not. It's not relevant if we need to drain all skipped candidates anyway (node decommission), but is relevant for balancing, where we don't.

Reply:
I am thinking that this imbalance caught by test will only be fixed with global rebalancing then. test_load_balancing_with_random_load is essentially disabled today, and perhaps should remain that way until we come up with a fix.

Reply:
@tgrabiec why isn't the internode plan done on a per rack basis instead of dc? that's for handling a transition where racks are being added and we want to reduce the load of preexisting ones?

Reply:
> @tgrabiec why isn't the internode plan done on a per rack basis instead of dc? that's for handling a transition where racks are being added and we want to reduce the load of preexisting ones?

In order to be able to migrate tablets across racks. Adding a rack is one such scenario.